lightning/src/ln/channelmanager.rs

```diff
@@ -3684,73 +3779,95 @@ where
+								log_given_level!(self.logger, if e == ChannelMonitorUpdateStatus::PermanentFailure { Level::Error } else { Level::Info },
+									"Failed to update channel monitor with preimage {:?} immediately prior to force-close: {:?}",
 									payment_preimage, e);
-								let err = handle_monitor_update_res!(self, e, chan, RAACommitmentOrder::CommitmentFirst, false, msgs.is_some()).unwrap_err();
-								mem::drop(channel_state_lock);
-								self.handle_monitor_update_completion_actions(completion_action(Some(htlc_value_msat)));
-								return Err((counterparty_node_id, err));
-							}
+							},
 						}
-						if let Some((msg, commitment_signed)) = msgs {
-							log_debug!(self.logger, "Claiming funds for HTLC with preimage {} resulted in a commitment_signed for channel {}",
-								log_bytes!(payment_preimage.0), log_bytes!(chan.get().channel_id()));
-							channel_state.pending_msg_events.push(events::MessageSendEvent::UpdateHTLCs {
-								node_id: chan.get().get_counterparty_node_id(),
-								updates: msgs::CommitmentUpdate {
-									update_add_htlcs: Vec::new(),
-									update_fulfill_htlcs: vec![msg],
-									update_fail_htlcs: Vec::new(),
-									update_fail_malformed_htlcs: Vec::new(),
-									update_fee: None,
-									commitment_signed,
-								}
-							});
+						let (drop, res) = convert_chan_err!(self, e, chan.get_mut(), &chan_id);
+						if drop {
+							chan.remove_entry();
 						}
-						mem::drop(channel_state_lock);
-						self.handle_monitor_update_completion_actions(completion_action(Some(htlc_value_msat)));
-						Ok(())
-					} else {
-						Ok(())
-					}
-				},
-				Err((e, monitor_update)) => {
-					match self.chain_monitor.update_channel(chan.get().get_funding_txo().unwrap(), monitor_update) {
-						ChannelMonitorUpdateStatus::Completed => {},
-						e => {
-							// TODO: This needs to be handled somehow - if we receive a monitor update
-							// with a preimage we *must* somehow manage to propagate it to the upstream
-							// channel, or we must have an ability to receive the same update and try
-							// again on restart.
-							log_given_level!(self.logger, if e == ChannelMonitorUpdateStatus::PermanentFailure { Level::Error } else { Level::Info },
-								"Failed to update channel monitor with preimage {:?} immediately prior to force-close: {:?}",
-								payment_preimage, e);
-						},
-					}
-					let (drop, res) = convert_chan_err!(self, e, chan.get_mut(), &chan_id);
-					if drop {
-						chan.remove_entry();
-					}
-					mem::drop(channel_state_lock);
-					self.handle_monitor_update_completion_actions(completion_action(None));
-					Err((counterparty_node_id, res))
-				},
+						mem::drop(peer_state_opt);
+						mem::drop(per_peer_state_lock);
+						self.handle_monitor_update_completion_actions(completion_action(None));
+						Err((counterparty_node_id, res))
+					},
+				}
+			} else {
+				// We've held the peer_state mutex since finding the channel and setting
```

Comment:
Hmm this is true even with the only the `read` lock acquired unless I'm missing something, no?

Only one thread at a time can acquire the inner `peer_state` mutex for the same peer, if two or more threads with the `read` lock lock acquired is accessing the same peer's inner `peer_state`. Since we've been holding the inner `peer_state` mutex acquired in the `peer_state_opt` when finding the channel, no other thread can have acquired the inner `peer_state` mutex and removed the channel.
And since we have the `read` lock acquired, no other thread can acquire the `write` lock at the same time and remove the entire peer entry in the outer `per_peer_state` RwLock.

Let me know if I'm missing something :).